### ISSUE TITLE
Custom implementation of `fromJson` no longer generated `fromJson`

### ISSUE THREAD

**Issue by petermnt** [rrousselGit/freezed#285]
**Describe the bug**
Since #283 the generation no longer generates the `fromJson` code when there is an implementation of the `fromJson` factory method

**To Reproduce**
```dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'test.freezed.dart';
part 'test.g.dart';

@freezed
abstract class Model with _$Model {
  const factory Model(String a) = _Model;

  factory Model.fromJson(Map<String, dynamic> json) => _$ModelFromJson(json);
}

@freezed
abstract class Model2 with _$Model2 {
  const factory Model2(String a) = _Model2;

  factory Model2.fromJson(Map<String, dynamic> json) {
    print("Any custom code breaks generation");
    return _$Model2FromJson(json);
  }
}
```

**Expected behavior**
Generation should also happen in the case where there isn't a redirecting constructor.

**Comment by rrousselGit** [comment]
This case was never voluntarily supported

**Comment by petermnt** [comment]
The reason we use it is because we have a 'fallback' type. Maybe this can be supported in another way?

```dart
  factory Composition.fromJson(Map<String, dynamic> json) {
    try {
      return _$CompositionFromJson(json);
      // ignore: avoid_catching_errors
    } on FallThroughError {
      return _$_$FallbackCompositionFromJson(json);
    }
  }
```

**Comment by rrousselGit** [comment]
Maybe make a separate constructor?

```dart
factory Composition.example(Map<String, dynamic> json) {
     try {
      return Composition.fromJson(json);
      // ignore: avoid_catching_errors
    } on FallThroughError {
      return Fallback(json);
    }
}
```

**Comment by petermnt** [comment]
`Composition` is being used as a child of other `freezed` classes which is deserialized from json. 
Overriding the `fromJson` makes it that it automatically goes through our 'custom' code. With a separate constructor it wouldn't work like that.

**Comment by rrousselGit** [comment]
I see
Could you expand on the use-case?

**Comment by petermnt** [comment]
It's a bit complex, but I'll give it a shot.

We have a list with many view types (all named `XyzComposition`), using a `runtimeType` union key.
Most of these types have some common parameters such as 'title', 'uri' and 'backgroundColor'

We add types all the time, so for backwards compatibility we have a `FallbackComposition`. If a type is not supported by the version of the app, we parse it as `FallbackComposition` with the common fields as optional fields.

In the widget we build for that screen we then handle the nullable fields `title`, `uri` and `backgroundColor`.
We use these fields, and then also have a "please update the app" message added to the widget.


So in short, we have many union types, but don't want the app to break when we update our api and offer a graceful fallback.

**Comment by zulfahmi93** [comment]
I'm not whether I should report this here, or create a new issue. If I define the `fromJson` constructor using block body syntax, freezed will not generate the `fromJson` constructor.

This will not generate the `fromJson` constructor:
```
factory Address.fromJson(Map<String, dynamic> json) {
  return _$AddressFromJson(json);
}
```

However, this will:
`factory Address.fromJson(Map<String, dynamic> json) => _$AddressFromJson(json);`

**Comment by danielmahon** [comment]
I also use custom `fromJson` blocks to normalize API data that I may not have total control over. Before, it was trivial to convert keys to camelCase, for example. Especially when certain graphql/api clients already use the "standard" `fromJson/toJson` methods for serialization. Could indeed probably use custom constructors but wouldn't be as intuitive and would require redirecting the `fromJson` constructor all the way through nested models.

**Comment by scopendo** [comment]
I have also come across this issue.  In my case, the reason for the custom fromJson method is to return null if the input is null.

```dart
@freezed
abstract class Model implements _$Model {
  const Model._();

  const factory Model({
    String id,
    String name,
  }) = _Model;

  factory Model.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;
    return _$ModelFromJson(json);
  }
}
```

**Comment by rrousselGit** [comment]
Note that any factory with an arrow with work

So you can do:

```dart
factory Foo.fromJson(json) => function(json);


Foo function(json) {}
```

**Comment by petermnt** [comment]
> Note that any factory with an arrow with work
 
That is indeed a working workaround for us.

```dart
  factory Composition.fromJson(Map<String, dynamic> json) => Composition._fromJsonWithFallback(json);

  factory Composition._fromJsonWithFallback(Map<String, dynamic> json) {
    try {
      return _$CompositionFromJson(json);
      // ignore: avoid_catching_errors
    } on FallThroughError {
      return _$_$FallbackCompositionFromJson(json);
    }
  }
```

It's a bit of an intricacy that is hard to catch. I think a little addition to the description of the `fromJson` constructor the docs could clear this up for other people. But for use the workaround is good enough.

**Comment by QonvexTech** [comment]
just add part "<model_name>.g.dart";
it generated the lost method missing in <model_name>.freezed.dart

**Comment by rrousselGit** [comment]
You should now be able to do:

```dart
@Freezed(fromJson: true, toJson: true)
abstract class Model2 with _$Model2 {
  const factory Model2(String a) = _Model2;

  factory Model2.fromJson(Map<String, dynamic> json) {
    print("This is fine");
    return _$Model2FromJson(json);
  }
}
```
